File: packages/guards/src/resident-id.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface ResidentIdOptions {
  action: 'block' | 'warn' | 'mask';
  validateChecksum?: boolean;
}

interface RidMatch {
  value: string;
  start: number;
  end: number;
}

const RID_PATTERN = /\d{6}-[1-8]\d{6}/g;
const CHECKSUM_WEIGHTS = [2, 3, 4, 5, 6, 7, 8, 9, 2, 3, 4, 5];

function validateChecksum(raw: string): boolean {
  const digits = raw.replace('-', '');
  if (digits.length !== 13) return false;
  let sum = 0;
  for (let i = 0; i < 12; i++) {
    sum += Number(digits[i]) * CHECKSUM_WEIGHTS[i];
  }
  const check = (11 - (sum % 11)) % 10;
  return check === Number(digits[12]);
}

function detect(
  text: string,
  useChecksum: boolean,
): RidMatch[] {
  const matches: RidMatch[] = [];
  const re = new RegExp(RID_PATTERN.source, 'g');
  let m: RegExpExecArray | null;
  while ((m = re.exec(text)) !== null) {
    if (useChecksum && !validateChecksum(m[0])) continue;
    matches.push({
      value: m[0],
      start: m.index,
      end: m.index + m[0].length,
    });
  }
  return matches.sort((a, b) => b.start - a.start);
}

function maskRid(rid: string): string {
  const [front, back] = rid.split('-');
  return `${front}-${back[0]}******`;
}

function maskText(text: string, matches: RidMatch[]): string {
  let result = text;
  for (const m of matches) {
    result =
      result.slice(0, m.start) +
      maskRid(m.value) +
      result.slice(m.end);
  }
  return result;
}

export function residentId(options: ResidentIdOptions): Guard {
  const useChecksum = options.validateChecksum ?? true;

  return {
    name: 'resident-id',
    version: '0.1.0',
    description: 'Korean resident ID detection with checksum',
    category: 'privacy',
    supportedStages: ['input', 'output'],
    async check(
      text: string,
      _ctx: GuardContext,
    ): Promise<GuardResult> {
      const start = performance.now();
      const matches = detect(text, useChecksum);
      const triggered = matches.length > 0;

      if (!triggered) {
        return {
          guardName: 'resident-id',
          passed: true,
          action: 'allow',
          latencyMs: Math.round(performance.now() - start),
        };
      }

      if (options.action === 'mask') {
        return {
          guardName: 'resident-id',
          passed: true,
          action: 'override',
          overrideText: maskText(text, matches),
          latencyMs: Math.round(performance.now() - start),
          details: {
            detected: matches.map((m) => m.value),
          },
        };
      }

      return {
        guardName: 'resident-id',
        passed: false,
        action: options.action,
        latencyMs: Math.round(performance.now() - start),
        details: {
          detected: matches.map((m) => m.value),
        },
      };
    },
  };
}

File: python/open_guardrail/guards/resident_id.py
"""Generic resident/national ID detection guard."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "ssn": re.compile(r"\d{3}-\d{2}-\d{4}"),
    "national-id": re.compile(r"\d{9,14}"),
}

_MASK_LABELS: dict[str, str] = {
    "ssn": "[SSN]",
    "national-id": "[NATIONAL-ID]",
}


def _detect(
    text: str, entities: list[str]
) -> list[dict]:
    matches: list[dict] = []
    for entity in entities:
        pat = _PATTERNS.get(entity)
        if not pat:
            continue
        for m in pat.finditer(text):
            matches.append(
                {
                    "type": entity,
                    "value": m.group(),
                    "start": m.start(),
                    "end": m.end(),
                }
            )
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask_text(text: str, matches: list[dict]) -> str:
    result = text
    for m in matches:
        label = _MASK_LABELS.get(m["type"], "[PII]")
        result = (
            result[: m["start"]] + label + result[m["end"] :]
        )
    return result


class _ResidentId:
    def __init__(
        self,
        *,
        action: str = "block",
        entities: Optional[List[str]] = None,
    ) -> None:
        self.name = "resident-id"
        self.action = action
        self._entities = entities or list(_PATTERNS.keys())

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._entities)
        elapsed = (time.perf_counter() - start) * 1000

        if not matches:
            return GuardResult(
                guard_name="resident-id",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )

        detected = [
            {"type": m["type"], "value": m["value"]}
            for m in matches
        ]

        if self.action == "mask":
            return GuardResult(
                guard_name="resident-id",
                passed=True,
                action="override",
                override_text=_mask_text(text, matches),
                latency_ms=round(elapsed, 2),
                details={"detected": detected},
            )

        return GuardResult(
            guard_name="resident-id",
            passed=False,
            action=self.action,
            message="Resident ID detected in text",
            latency_ms=round(elapsed, 2),
            details={
                "detected": detected,
                "reason": (
                    "Text contains resident or national"
                    " identification numbers"
                ),
            },
        )


def resident_id(
    *,
    action: str = "block",
    entities: Optional[List[str]] = None,
) -> _ResidentId:
    return _ResidentId(action=action, entities=entities)
